### PR TITLE
Add support for IPv6 in WAMR

### PR DESCRIPTION
**Pull request by loganek**
For now this implementation only covers posix platforms, as defined in MVP #1336

### CODE REVIEW

File: core/shared/platform/common/posix/posix_socket.c [hunk 1]
```diff
@@ -9,28 +9,57 @@
...
+        v6->sin6_port = htons(port);
+        return true;
+    }
+
+    return false;
 }
 
 int
-os_socket_create(bh_socket_t *sock, int tcp_or_udp)
+os_socket_create(bh_socket_t *sock, int ip4_or_ip6, int tcp_or_udp)
```
Comment (wenyongh) [inline core/shared/platform/common/posix/posix_socket.c:38]: The parameter definitions make a little confused, how about change to:
`bool is_ipv4, bool is_tcp`? And add comments in the API declaration.

File: core/shared/platform/common/posix/posix_socket.c [hunk 2]
```diff
@@ -166,13 +201,28 @@ os_socket_shutdown(bh_socket_t socket)
 }
 
 int
-os_socket_inet_network(const char *cp, uint32 *out)
+os_socket_inet_network(uint8 ipv4_or_ipv6, const char *cp, uint8 *out)
```
Comment (wenyongh) [inline core/shared/platform/common/posix/posix_socket.c:204]: Same as above, how about `bool is_ipv4`?
And for `uint8 *out`, there may be address unaligned issue, if the input is not 4-byte aligned, L218, `*ipv4_buffer = ntohl(*ipv4_buffer);` may cause core dump in some targets, e.g. arm?

How about define an union type, like:
```C
union {
    uint32 ipv4;
    uint16 ipv6[8];
};
```

File: core/iwasm/libraries/lib-socket/src/wasi/wasi_socket_ext.c [hunk 3]
```diff
@@ -31,21 +31,38 @@ ipv4_addr_to_wasi_addr(uint32_t addr_num, uint16_t port, __wasi_addr_t *out)
...
+    out->kind = IPv6;
+    out->addr.ip6.port = ntohs(port);
+    out->addr.ip6.addr.n0 = ntohs(addr[0]);
+    out->addr.ip6.addr.n1 = ntohs(addr[1]);
+    out->addr.ip6.addr.n2 = ntohs(addr[2]);
+    out->addr.ip6.addr.n3 = ntohs(addr[3]);
+    out->addr.ip6.addr.h0 = ntohs(addr[4]);
+    out->addr.ip6.addr.h1 = ntohs(addr[5]);
+    out->addr.ip6.addr.h2 = ntohs(addr[6]);
+    out->addr.ip6.addr.h3 = ntohs(addr[7]);
```
Comment (wenyongh) [inline core/iwasm/libraries/lib-socket/src/wasi/wasi_socket_ext.c:46]: Found that there are some similar conversions, including PR #1413 [wasi_socket_ext.c](https://github.com/bytecodealliance/wasm-micro-runtime/pull/1413/files#diff-798a0826792097576b78eaa27018727f43ecae36c7296bdafada5bf4c614b54e), shall we add common APIs for the conversions?

File: core/shared/platform/common/posix/posix_socket.c (hunk 2 above)
Comment (loganek) [reply core/shared/platform/common/posix/posix_socket.c]: Good point, updated.

File: core/iwasm/libraries/lib-socket/src/wasi/wasi_socket_ext.c (hunk 3 above)
Comment (loganek) [reply core/iwasm/libraries/lib-socket/src/wasi/wasi_socket_ext.c:46]: this is a copy-paste from that file. Once we merge one of the PRs, we can remove this part from the second. Just wanted to have both PRs working.

**Comment by wenyongh** [comment]
LGTM

File: core/iwasm/libraries/lib-socket/src/wasi/wasi_socket_ext.c [hunk 4]
```diff
@@ -31,21 +31,38 @@ ipv4_addr_to_wasi_addr(uint32_t addr_num, uint16_t port, __wasi_addr_t *out)
...
+        assert(sizeof(struct sockaddr_in) <= addrlen);
 
         ipv4_addr_to_wasi_addr(
             ((struct sockaddr_in *)sock_addr)->sin_addr.s_addr,
             ((struct sockaddr_in *)sock_addr)->sin_port, wasi_addr);
     }
     else if (AF_INET6 == sock_addr->sa_family) {
-        // TODO: IPV6
-        ret = __WASI_ERRNO_AFNOSUPPORT;
+        assert(sizeof(struct sockaddr_in6) <= addrlen);
```
Comment (lum1n0us) [inline core/iwasm/libraries/lib-socket/src/wasi/wasi_socket_ext.c:62]: IIUC, `sock_addr` should be either a value of `struct sockaddr_in` or `struct sockaddr_in6`. So, L55 using `<=` is right. But L62 should use `==`. There is no third option.

**Review by lum1n0us — APPROVED** [review]
Just one little suggestion.

File: core/iwasm/libraries/lib-socket/src/wasi/wasi_socket_ext.c (hunk 4 above)
Comment (loganek) [reply core/iwasm/libraries/lib-socket/src/wasi/wasi_socket_ext.c]: done

**Comment by lum1n0us** [comment]
LGTM

**Pull request merged**